packages/preferences/src/browser/preferences-tree-widget.ts

```diff
@@ -236,36 +289,116 @@ export class PreferencesEditorsContainer extends DockPanel {
+        if (editor) {
+            this.onEditorChangedEmitter.fire(<PreferencesEditorWidget>editor);
+        }
         super.onUpdateRequest(msg);
     }
 
     protected async onAfterAttach(msg: Message): Promise<void> {
+        this._userPreferenceEditorWidget = await this.getUserPreferenceEditorWidget();
+        this.addWidget(this._userPreferenceEditorWidget);
+        await this.refreshWorkspacePreferenceEditor();
+        await this.refreshFoldersPreferencesEditorWidget(undefined);
+
+        super.onAfterAttach(msg);
+        this.onInitEmitter.fire(undefined);
+    }
+
+    protected async getUserPreferenceEditorWidget(): Promise<PreferencesEditorWidget> {
         const userPreferenceUri = this.userPreferenceProvider.getUri();
         const userPreferences = await this.editorManager.getOrCreateByUri(userPreferenceUri) as PreferencesEditorWidget;
-        userPreferences.title.label = 'User Preferences';
-        userPreferences.title.caption = await this.getPreferenceEditorCaption(userPreferenceUri);
+        userPreferences.title.label = 'User';
+        userPreferences.title.caption = `User Preferneces: ${await this.getPreferenceEditorCaption(userPreferenceUri)}`;
         userPreferences.scope = PreferenceScope.User;
-        this.addWidget(userPreferences);
+        return userPreferences;
+    }
+
+    async refreshWorkspacePreferenceEditor(): Promise<void> {
+        const newWorkspacePreferenceEditorWidget = await this.getWorkspacePreferenceEditorWidget();
+        if (newWorkspacePreferenceEditorWidget) {
+            this.addWidget(newWorkspacePreferenceEditorWidget,
+                { ref: this._workspacePreferenceEditorWidget || this._userPreferenceEditorWidget });
+            if (this._workspacePreferenceEditorWidget) {
+                this._workspacePreferenceEditorWidget.close();
+                this._workspacePreferenceEditorWidget.dispose();
+            }
+            this._workspacePreferenceEditorWidget = newWorkspacePreferenceEditorWidget;
+        }
+    }
 
+    protected async getWorkspacePreferenceEditorWidget(): Promise<PreferencesEditorWidget | undefined> {
         const workspacePreferenceUri = await this.workspacePreferenceProvider.getUri();
         const workspacePreferences = workspacePreferenceUri && await this.editorManager.getOrCreateByUri(workspacePreferenceUri) as PreferencesEditorWidget;
 
         if (workspacePreferences) {
-            workspacePreferences.title.label = 'Workspace Preferences';
-            workspacePreferences.title.caption = await this.getPreferenceEditorCaption(workspacePreferenceUri!);
+            workspacePreferences.title.label = 'Workspace';
+            workspacePreferences.title.caption = `Workspace Preferences: ${await this.getPreferenceEditorCaption(workspacePreferenceUri!)}`;
+            workspacePreferences.title.icon = 'database-icon medium-yellow file-icon';
+            workspacePreferences.editor.setLanguage(JSONC_LANGUAGE_ID);
```

Comment:
We add the "language" setting  here for the workspace editor, but we didn't add it for the user and the folder workspaces

Reply:
it is because the user and folder settings files use `.json` file extension - it can be recognised by theia, while workspace settings file (for now) could possibly have any extension. We don't force users to use `theia-workspace` or `code-workspace` - maybe we should ?